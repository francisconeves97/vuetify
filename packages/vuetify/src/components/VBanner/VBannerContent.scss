// Imports
@import './index';

.v-banner__content {
  align-items: center;
  display: flex;
  flex: 1 1 auto;
  line-height: $banner-content-line-height;
  padding-bottom: $banner-content-padding-bottom;

  @include banner-mobile () {
    padding-bottom: $banner-mobile-content-padding-bottom;
  }

  @include banner-single-line () {
    padding-bottom: $banner-single-line-content-padding-bottom;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }

  @include banner-thumbnail () {
    padding-bottom: $banner-mobile-content-padding-bottom;
  }
}
