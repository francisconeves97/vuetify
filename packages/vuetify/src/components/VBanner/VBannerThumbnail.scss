// Imports
@import './index';

.v-banner-thumbnail {
  flex: 0 1 auto;

  @include rounded($banner-query);

  @include ltr () {
    margin-left: $banner-thumbnail-margin-start;
    margin-right: $banner-thumbnail-margin-end;
  }

  @include rtl () {
    margin-right: $banner-thumbnail-margin-start;
    margin-left: $banner-thumbnail-margin-end;
  }

  &--is-mobile {
    @include ltr() {
      margin-right: $banner-mobile-thumbnail-padding-start;
    }

    @include rtl() {
      margin-left: $banner-mobile-thumbnail-padding-start;
    }
  }
}

.v-banner-thumbnail__avatar {
  border-radius: $banner-avatar-border-radius;
}
