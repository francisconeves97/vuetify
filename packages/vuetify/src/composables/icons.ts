// Utilities
import { computed, h, inject } from 'vue'

// Components
import { VComponentIcon } from '@/components'

// Types
import type { Component, InjectionKey, Ref, VNode } from 'vue'

export type IconValue = string | Component

export interface IconAliases {
  [name: string]: IconValue
  complete: IconValue
  cancel: IconValue
  close: IconValue
  delete: IconValue
  clear: IconValue
  success: IconValue
  info: IconValue
  warning: IconValue
  error: IconValue
  prev: IconValue
  next: IconValue
  checkboxOn: IconValue
  checkboxOff: IconValue
  checkboxIndeterminate: IconValue
  delimiter: IconValue
  sort: IconValue
  expand: IconValue
  menu: IconValue
  subgroup: IconValue
  dropdown: IconValue
  radioOn: IconValue
  radioOff: IconValue
  edit: IconValue
  ratingEmpty: IconValue
  ratingFull: IconValue
  ratingHalf: IconValue
  loading: IconValue
  first: IconValue
  last: IconValue
  unfold: IconValue
  file: IconValue
  plus: IconValue
  minus: IconValue
}

export interface IconProps {
  tag: string
  icon: IconValue
  disabled?: Boolean
  class?: unknown[]
  style?: Record<string, unknown> | null
  'aria-hidden': boolean
  type?: string
}

export interface IconSet {
  component: (props: IconProps) => VNode
}

export type IconOptions = {
  defaultSet: string
  aliases?: Partial<IconAliases>
  sets: Record<string, IconSet>
}

type IconInstance = {
  component: (props: IconProps) => VNode
  icon: IconValue
}

export const VuetifyIconSymbol: InjectionKey<IconOptions> = Symbol.for('vuetify:icons')

export const useIcon = (props: { icon: IconValue }) => {
  const icons = inject(VuetifyIconSymbol)

  if (!icons) throw new Error('Missing Vuetify Icons provide!')

  const icon: Ref<IconInstance> = computed(() => {
    let icon: IconValue | undefined = props.icon

    if (typeof props.icon === 'string' && props.icon.includes('$')) {
      icon = icons.aliases?.[props.icon.slice(props.icon.indexOf('$') + 1)]
    }

    if (!icon) throw new Error(`Could not find aliased icon ${props.icon}`)

    if (typeof icon !== 'string') {
      return {
        component: iconProps => h(VComponentIcon, iconProps as any), // TODO: fix any
        icon,
      }
    }

    const hasSet = icon.includes(':')
    const setName = hasSet ? icon.split(':')[0] : icons.defaultSet
    const iconName = hasSet ? icon.split(':')[1] : icon
    const set = icons.sets[setName ?? icons.defaultSet]

    if (!set) {
      // TODO: Throw error?
      return {
        component: () => h('div', ['error!']),
        icon: iconName,
      }
    }

    return {
      component: set.component,
      icon: iconName,
    }
  })

  return { icon }
}
