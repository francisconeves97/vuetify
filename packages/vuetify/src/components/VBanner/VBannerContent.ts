// Styles
import './VBannerContent.scss'

// Components
import VBannerThumbnail, { makeThumbnailProps } from './VBannerThumbnail'

// Utilities
import { defineComponent, h } from 'vue'
import makeProps from '@/util/makeProps'

export default defineComponent({
  name: 'VBannerContent',

  inheritAttrs: false,

  props: makeProps(makeThumbnailProps()),

  setup (props, { slots }) {
    return () => h('div', {
      class: 'v-banner__content',
    }, [
      (!!props.avatar || !!props.icon || !!slots.thumbnail) && h(VBannerThumbnail, props, slots),
      slots.default?.(),
    ])
  },
})
