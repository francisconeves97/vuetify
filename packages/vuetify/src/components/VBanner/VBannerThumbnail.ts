// Styles
import './VBannerThumbnail.scss'

// Utilities
import { defineComponent, h } from 'vue'
import makeProps from '@/util/makeProps'
import propsFactory from '@/util/propsFactory'

// Composables
export const makeThumbnailProps = propsFactory({
  avatar: String,
  icon: String,
})

export default defineComponent({
  name: 'VBannerThumbnail',

  inheritAttrs: false,

  props: makeProps(makeThumbnailProps()),

  render () {
    const children = []

    if (this.$slots.thumbnail) {
      children.push(this.$slots.thumbnail?.())
    } else if (this.$props.avatar) {
      children.push(
        h('div', {
          class: 'v-banner__thumbnail__avatar',
          style: {
            height: '36px',
            width: '36px',
          },
        })
      )
    }

    return h('div', {
      class: 'v-banner__thumbnail',
    }, children)
  },
})
