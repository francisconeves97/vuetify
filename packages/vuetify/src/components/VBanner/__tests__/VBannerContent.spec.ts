// Components
import VBannerContent from '../VBannerContent'
import VBannerThumbnail from '../VBannerThumbnail'

// Utilities
import { createTheme, VuetifyThemeSymbol } from '@/composables/theme'
import { mount } from '@vue/test-utils'
import { VuetifySymbol } from '@/framework'

describe('VBannerContent', () => {
  function mountFunction (options = {}) {
    return mount(VBannerContent, {
      global: {
        provide: {
          [VuetifySymbol as symbol]: { defaults: { global: {} } },
          [VuetifyThemeSymbol as symbol]: createTheme(),
        },
      },
      ...options,
    })
  }

  it('should match a snapshot', () => {
    const wrapper = mountFunction()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should generate thumbnail slot', () => {
    const wrapper = mountFunction({
      slots: { thumbnail: '<div>foobar</div>' },
    })

    expect(wrapper.html()).toContain('<div>foobar</div>')
  })

  it.each([
    [{}, false],
    [{ avatar: 'foobar' }, true],
    [{ icon: 'foobar' }, true],
  ])('should generate actions slot', (props, expected) => {
    const wrapper = mountFunction({ props })
    const thumbnail = wrapper.findComponent(VBannerThumbnail)

    expect(thumbnail.exists()).toBe(expected)
  })
})
