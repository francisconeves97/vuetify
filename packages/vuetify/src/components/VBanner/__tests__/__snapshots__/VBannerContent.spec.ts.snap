// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VBannerContent should match a snapshot 1`] = `
<div class="v-banner__content">
</div>
`;
